examples/routing_service/routing_service_file_adapter/cpp/FileInputDiscoveryStreamReader.cxx

```diff
@@ -0,0 +1,108 @@
+{
+    std::ifstream input_file;
+    input_file.open(filename);
+    return input_file.is_open();
+}
+
+FileInputDiscoveryStreamReader::FileInputDiscoveryStreamReader( 
+        const PropertySet &, 
+        StreamReaderListener *input_stream_discovery_listener)
+        : data_samples_mutex_(), 
+        data_samples_()
+{
+    input_stream_discovery_listener_ = input_stream_discovery_listener;
+
+    /**
+     * In our example, we provide statically the stream information available. 
+     * We do not have a mechanism demonstrating how to perform discovery after 
+     * startup. However, as an idea you can have a thread monitoring the file 
+     * system and updating the list of StreamInfo samples and calling 
+     * input_stream_discovery_listener_->on_data_available(this); to notify that 
+     * new files have been discovered.
+     */
+    if (fexists(SQUARE_FILE_NAME)) {
+        this->data_samples_.push_back(new StreamInfo("Square", "ShapeType"));
+    }
+
+    if (fexists(CIRCLE_FILE_NAME)) {
+        this->data_samples_.push_back(new StreamInfo("Circle", "ShapeType"));
+    }
+
+    if (fexists(TRIANGLE_FILE_NAME)) {
+        this->data_samples_.push_back(new StreamInfo("Triangle", "ShapeType"));
+    }
+
+    /** 
+     * Once the FileInputDiscoveryStreamReader is initialized, we trigger an 
+     * event to notify that the streams are ready.
+     */
+    input_stream_discovery_listener_->on_data_available(this);
+}
+
+void FileInputDiscoveryStreamReader::dispose(
+        const rti::routing::StreamInfo &stream_info)
+{
+    /**
+     * This guard is essential since the take() and return_loan() operations triggered 
+     * by calling on_data_available() execute on a different thread. The take() and 
+     * return_loan() operations also need to access the data_samples_ list.
+     */
+    std::lock_guard<std::mutex> guard(data_samples_mutex_);
+
+    rti::routing::StreamInfo *stream_info_disposed = new StreamInfo(
+            stream_info.stream_name(), 
+            stream_info.type_info().type_name());
+    stream_info_disposed->disposed(true);
+
+    this->data_samples_.push_back(stream_info_disposed);
+    input_stream_discovery_listener_->on_data_available(this);
+}
+
+void FileInputDiscoveryStreamReader::take(
+        std::vector<rti::routing::StreamInfo *> &stream)
+{
+    /**
+     * This guard is essential since the take() and return_loan() operations triggered 
+     * by calling on_data_available() execute on a different thread. The take() and 
+     * return_loan() operations also need to access the data_samples_ list.
+     */
+    std::lock_guard<std::mutex> guard(data_samples_mutex_);
+    std::copy(data_samples_.begin(), data_samples_.end(), std::back_inserter(stream));
+}
+
+void FileInputDiscoveryStreamReader::return_loan(
+        std::vector<rti::routing::StreamInfo *> &stream)
+{
+    /**
+     * This guard is essential since the take() and return_loan() operations triggered 
```

Comment:
Why do take() and return_loan() run on different threads? Shouldn't RS call:

take()
...
return_loan()

In the same thread?

Reply:
Yes it does. Maybe my comment didn't convey the idea clearly. I meant dispose() doesn't run on the same thread as take() or return_loan(). I have updated the comment to be more clear.

Reply:
take() and return_loan() runs in the same thread, but the function that adds sample to the queue is in a separate thread (The one that reads the files). So that's why the mutex is needed